coderd/members.go

```diff
@@ -87,30 +94,94 @@ func (api *API) putMemberRoles(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	httpapi.Write(ctx, rw, http.StatusOK, convertOrganizationMember(updatedUser))
+	resp := convertOrganizationMembers(ctx, api.Database, []database.OrganizationMember{updatedUser})
+	if len(resp) != 1 {
+		httpapi.InternalServerError(rw, xerrors.Errorf("failed to serialize member to response, update still succeeded"))
+		return
+	}
+	httpapi.Write(ctx, rw, http.StatusOK, resp[0])
 }
 
-func convertOrganizationMember(mem database.OrganizationMember) codersdk.OrganizationMember {
-	convertedMember := codersdk.OrganizationMember{
-		UserID:         mem.UserID,
-		OrganizationID: mem.OrganizationID,
-		CreatedAt:      mem.CreatedAt,
-		UpdatedAt:      mem.UpdatedAt,
-		Roles:          make([]codersdk.SlimRole, 0, len(mem.Roles)),
+// convertOrganizationMembers batches the role lookup to make only 1 sql call
+// We
+func convertOrganizationMembers(ctx context.Context, db database.Store, mems []database.OrganizationMember) []codersdk.OrganizationMember {
+	converted := make([]codersdk.OrganizationMember, 0, len(mems))
+	roleLookup := make([]database.NameOrganizationPair, 0)
+
+	for _, m := range mems {
+		converted = append(converted, codersdk.OrganizationMember{
+			UserID:         m.UserID,
+			OrganizationID: m.OrganizationID,
+			CreatedAt:      m.CreatedAt,
+			UpdatedAt:      m.UpdatedAt,
+			Roles: db2sdk.List(m.Roles, func(r string) codersdk.SlimRole {
+				// If it is a built-in role, no lookups are needed.
+				rbacRole, err := rbac.RoleByName(rbac.RoleIdentifier{Name: r, OrganizationID: m.OrganizationID})
+				if err == nil {
+					return db2sdk.SlimRole(rbacRole)
+				}
+
+				// We know the role name and the organization ID. We are missing the
+				// display name. Append the lookup parameter, so we can get the display name
+				roleLookup = append(roleLookup, database.NameOrganizationPair{
+					Name:           r,
+					OrganizationID: m.OrganizationID,
+				})
+				return codersdk.SlimRole{
+					Name:           r,
+					DisplayName:    "",
+					OrganizationID: m.OrganizationID.String(),
+				}
+			}),
+		})
 	}
 
-	for _, roleName := range mem.Roles {
-		rbacRole, _ := rbac.RoleByName(rbac.RoleIdentifier{Name: roleName, OrganizationID: mem.OrganizationID})
-		convertedMember.Roles = append(convertedMember.Roles, db2sdk.SlimRole(rbacRole))
+	customRoles, err := db.CustomRoles(ctx, database.CustomRolesParams{
+		LookupRoles:     roleLookup,
+		ExcludeOrgRoles: false,
+		OrganizationID:  uuid.UUID{},
+	})
+	if err != nil {
+		// We are missing the display names, but that is not absolutely required. So just
```

Comment:
Are we expecting an error here? I see we _can_ do without it, but I wonder if we should instead of just presenting the error?

Reply:
I'll return the error :+1:

Reply:
Just unfortunate we can block api calls on this kind of error. Would be nice if we could return some sort of "warning" instead. I'll return an error though as that is what we usually do for failed db calls.